lib/demuxeres/gstdemuxeres.c

```diff
@@ -0,0 +1,721 @@
+  stream->stream_id = gst_pad_get_stream_id (GST_PAD_CAST (pad));
+  GstStructure *s = gst_caps_get_structure (caps, 0);
+  switch (stream->type) {
+    case DEMUXER_ES_STREAM_TYPE_VIDEO:
+    {
+      gst_video_info_from_caps (&stream->data.video.info, caps);
+      gst_structure_get_int (s, "bitrate", &stream->data.video.bitrate);
+      stream->data.video.profile =
+          g_strdup (gst_structure_get_string (s, "profile"));
+      stream->data.video.level =
+          g_strdup (gst_structure_get_string (s, "level"));
+      stream->data.video.vcodec = gst_parse_stream_get_vcodec_from_caps (caps);
+      break;
+    }
+    case DEMUXER_ES_STREAM_TYPE_AUDIO:
+    {
+      gst_audio_info_from_caps (&stream->data.audio.info, caps);
+      gst_structure_get_int (s, "bitrate", &stream->data.audio.bitrate);
+      stream->data.audio.acodec = gst_parse_stream_get_acodec_from_caps (caps);
+      break;
+    }
+    default:
+      break;
+  }
+  stream->id = g_list_length (priv->streams);
+
+  funnel_pad = gst_element_request_pad_simple (priv->funnel, "sink_%u");
+
+  if (gst_pad_link (pad, funnel_pad) != GST_PAD_LINK_OK)
+    GST_ERROR ("Unable to plug the pad %p to the capsfilter pad", pad);
+
+  gst_caps_unref (caps);
+
+  gst_object_unref (funnel_pad);
+
+  GST_DEBUG_BIN_TO_DOT_FILE_WITH_TS (GST_BIN (priv->pipeline),
+      GST_DEBUG_GRAPH_SHOW_ALL, "gst-demuxerer.stream_create");
+
+  return stream;
+}
+
+static inline gboolean
+wait_for_demuxer_ready (GstDemuxerES * demuxer)
+{
+  GstDemuxerESPrivate *priv = demuxer->priv;
+
+  g_mutex_lock (&priv->ready_mutex);
+  while (priv->state == DEMUXER_ES_STATE_IDLE) {
+    g_cond_wait (&priv->ready_cond, &priv->ready_mutex);
+  }
+  g_mutex_unlock (&priv->ready_mutex);
+
+  return (priv->state >= DEMUXER_ES_STATE_READY);
+}
+
+static void
+uridecodebin_pad_added_cb (GstElement * uridecodebin, GstPad * pad,
+    GstDemuxerES * demuxer)
+{
+  GstDemuxerEStream *stream = NULL;
+
+  if (!GST_PAD_IS_SRC (pad))
+    return;
+
+  GST_DEBUG ("pad %s:%s", GST_DEBUG_PAD_NAME (pad));
+
+  stream = gst_parse_stream_create (demuxer, pad);
+  if (!stream)
+    return;
+
+  demuxer->priv->streams = g_list_append (demuxer->priv->streams, stream);
+  GST_DEBUG ("Done linking");
+}
+
+static void
+uridecodebin_pad_no_more_pads (GstElement * uridecodebin,
+    GstDemuxerES * demuxer)
+{
+  GST_INFO ("No more pads received from %s", GST_ELEMENT_NAME (uridecodebin));
+  if (demuxer->priv->state == DEMUXER_ES_STATE_IDLE) {
+    set_demuxer_state (demuxer, DEMUXER_ES_STATE_READY);
+  }
+}
+
+static GstAutoplugSelectResult
+uridecodebin_autoplug_select_cb (GstElement * src, GstPad * pad,
+    GstCaps * caps, GstElementFactory * factory, GstDemuxerES * demuxer)
+{
+  GstAutoplugSelectResult ret = GST_AUTOPLUG_SELECT_TRY;
+  if (gst_element_factory_list_is_type (factory,
+          GST_ELEMENT_FACTORY_TYPE_DECODER)) {
+    GST_DEBUG ("Expose pad if factory is decoder.");
+    ret = GST_AUTOPLUG_SELECT_EXPOSE;
```

Comment:
I don't understand the need of this callback.

Reply:
This callback is necessary to tell  we want to stop before the decoder in idecodebin. Is there a better way to tell  it ?

Reply:
got it!